lib/src/tokenizer.dart

```diff
@@ -43,76 +43,76 @@ class HtmlTokenizer implements Iterator<Token> {
 
   /// Holds a reference to the method to be invoked for the next parser state.
-  bool Function() state;
+  late bool Function() state;
 
   final StringBuffer _buffer = StringBuffer();
 
-  int _lastOffset;
+  late int _lastOffset;
 
   // TODO(jmesserly): ideally this would be a LinkedHashMap and we wouldn't add
   // an item until it's ready. But the code doesn't have a clear notion of when
   // it's "done" with the attribute.
-  List<TagAttribute> _attributes;
-  Set<String> _attributeNames;
+  List<TagAttribute>? _attributes;
+  Set<String>? _attributeNames;
 
   HtmlTokenizer(doc,
-      {String encoding,
+      {String? encoding,
       bool parseMeta = true,
       this.lowercaseElementName = true,
       this.lowercaseAttrName = true,
       this.generateSpans = false,
-      String sourceUrl,
+      String? sourceUrl,
       this.attributeSpans = false})
       : stream =
             HtmlInputStream(doc, encoding, parseMeta, generateSpans, sourceUrl),
         tokenQueue = Queue() {
     reset();
   }
 
-  TagToken get currentTagToken => currentToken as TagToken;
-  DoctypeToken get currentDoctypeToken => currentToken as DoctypeToken;
-  StringToken get currentStringToken => currentToken as StringToken;
+  TagToken? get currentTagToken => currentToken as TagToken?;
+  DoctypeToken? get currentDoctypeToken => currentToken as DoctypeToken?;
+  StringToken? get currentStringToken => currentToken as StringToken?;
 
-  Token _current;
+  Token? _current;
   @override
-  Token get current => _current;
+  Token get current => _current!;
 
   final StringBuffer _attributeName = StringBuffer();
   final StringBuffer _attributeValue = StringBuffer();
 
   void _markAttributeEnd(int offset) {
-    _attributes.last.value = '$_attributeValue';
-    if (attributeSpans) _attributes.last.end = stream.position + offset;
+    _attributes!.last.value = '$_attributeValue';
+    if (attributeSpans) _attributes!.last.end = stream.position + offset;
   }
 
   void _markAttributeValueStart(int offset) {
-    if (attributeSpans) _attributes.last.startValue = stream.position + offset;
+    if (attributeSpans) _attributes!.last.startValue = stream.position + offset;
   }
 
   void _markAttributeValueEnd(int offset) {
-    if (attributeSpans) _attributes.last.endValue = stream.position + offset;
+    if (attributeSpans) _attributes!.last.endValue = stream.position + offset;
     _markAttributeEnd(offset);
   }
 
   // Note: we could track the name span here, if we need it.
   void _markAttributeNameEnd(int offset) => _markAttributeEnd(offset);
 
-  void _addAttribute(String name) {
+  void _addAttribute(String? name) {
     _attributes ??= [];
     _attributeName.clear();
     _attributeName.write(name);
     _attributeValue.clear();
     final attr = TagAttribute();
-    _attributes.add(attr);
-    if (attributeSpans) attr.start = stream.position - name.length;
+    _attributes!.add(attr);
+    if (attributeSpans) attr.start = stream.position - name!.length;
```

Comment:
If `name` cannot be `null`, make the parameter non-nullable.

Reply:
Done. The diff for this one overlaps some with the change to make `isLetter` take a non-nullable argument, so maybe that's some evidence that it would be worth doing.

```diff
diff --git a/lib/src/tokenizer.dart b/lib/src/tokenizer.dart
index 0a7f660..4f49cb1 100644
--- a/lib/src/tokenizer.dart
+++ b/lib/src/tokenizer.dart
@@ -105,14 +105,14 @@ class HtmlTokenizer implements Iterator<Token> {
   // Note: we could track the name span here, if we need it.
   void _markAttributeNameEnd(int offset) => _markAttributeEnd(offset);
 
-  void _addAttribute(String? name) {
+  void _addAttribute(String name) {
     _attributes ??= [];
     _attributeName.clear();
     _attributeName.write(name);
     _attributeValue.clear();
     final attr = TagAttribute();
     _attributes!.add(attr);
-    if (attributeSpans) attr.start = stream.position - name!.length;
+    if (attributeSpans) attr.start = stream.position - name.length;
   }
 
   /// This is where the magic happens.
@@ -1006,7 +1006,7 @@ class HtmlTokenizer implements Iterator<Token> {
     final data = stream.char();
     if (isWhitespace(data)) {
       stream.charsUntil(spaceCharacters, true);
-    } else if (isLetter(data)) {
+    } else if (data != null && isLetter(data)) {
       _addAttribute(data);
       state = attributeNameState;
     } else if (data == '>') {
@@ -1099,7 +1099,7 @@ class HtmlTokenizer implements Iterator<Token> {
       state = beforeAttributeValueState;
     } else if (data == '>') {
       emitCurrentToken();
-    } else if (isLetter(data)) {
+    } else if (data != null && isLetter(data)) {
       _addAttribute(data);
       state = attributeNameState;
     } else if (data == '/') {
```